infrastructure/src/main/java/org/corfudb/infrastructure/BatchProcessor.java

```diff
@@ -114,127 +102,152 @@ private void recordRunnable(Runnable fsyncRunnable, Optional<Timer> fsyncTimer)
+     * @param type The request type
+     * @param req  The request message
+     * @return     returns a future result for the request, if it expects one
+     */
+    public <T> CompletableFuture<T> addTask(@Nonnull Type type, @Nonnull RequestMsg req) {
+        BatchWriterOperation<T> op = new BatchWriterOperation<>(type, req);
+        operationsQueue.add(op);
+        return op.getFutureResult();
+    }
 
+    private void process() {
         if (!sync) {
             log.warn("batchWriteProcessor: writes configured to not sync with secondary storage");
         }
 
         try {
             BatchWriterOperation lastOp = null;
-            int processed = 0;
             List<BatchWriterOperation> res = new LinkedList<>();
+            int numProcessed = 0;
 
             while (true) {
-                BatchWriterOperation currOp;
+                BatchWriterOperation currentOp;
                 queueSizeDist.ifPresent(dist -> dist.record(operationsQueue.size()));
+
                 if (lastOp == null) {
-                    currOp = operationsQueue.take();
+                    currentOp = operationsQueue.take();
                 } else {
-                    currOp = operationsQueue.poll();
+                    currentOp = operationsQueue.poll();
 
-                    if (currOp == null || processed == BATCH_SIZE
-                            || currOp == BatchWriterOperation.SHUTDOWN) {
+                    if (currentOp == null || numProcessed == BATCH_SIZE || currentOp == BatchWriterOperation.SHUTDOWN) {
                         streamLog.sync(sync);
-                        log.trace("Completed {} operations", processed);
-
-                        for (BatchWriterOperation operation : res) {
-                            if (!operation.getFutureResult().isCompletedExceptionally()
-                            && !operation.getFutureResult().isCancelled()) {
-                                // At this point we need to complete the requests
-                                // that completed successfully (i.e. haven't failed)
-                                operation.getFutureResult().complete(operation.getResultValue());
+                        if (log.isTraceEnabled()) {
+                            log.trace("batchWriteProcessor: completed {} operations", numProcessed);
+                        }
+                        // At this point we need to complete the requests
+                        // that completed successfully (i.e. haven't failed)
+                        for (BatchWriterOperation op : res) {
+                            if (!op.getFutureResult().isCompletedExceptionally()
+                                    && !op.getFutureResult().isCancelled()) {
+                                op.getFutureResult().complete(op.getResultValue());
                             }
                         }
+
                         res.clear();
-                        processed = 0;
+                        numProcessed = 0;
```

Comment:
This loop is really messy and probably should be improved in another PR. 

The performance of this loop can be improved by using `ArrayList` instead of `LinkedList`. It produces less garbage and should reduce gc pressure.

Reply:
`numProcessed` can also be removed and replaced with `list.size()`

Reply:
I've replaced the `LinkedList` by an `ArrayList` and have removed `numProcessed`.